### PR TITLE
File Transfer RFD

### PR DESCRIPTION
**Pull request by ibeckermayer**
It took a lot of trial and error to whittle the scope down to this. There are several potential options at our disposal, each with highly detailed rabbit holes to go down to specify in full. I backed out of some initial attempts that would’ve needed to go into too much detail of each. This document is an RFD about which of those approaches to ultimately take, and it may demand further RFD’s to fully specify that chosen route.

### CODE REVIEW

File: rfd/0058-desktop-file-transfer.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+File transfer's are limited to files less than 4GB unless huge file support is enabled (`CB_HUGE_FILE_SUPPORT_ENABLED`).
```
Comment (zmb3) [inline rfd/0058-desktop-file-transfer.md:21]:
```suggestion
File transfers are limited to files less than 4GB unless huge file support is enabled (`CB_HUGE_FILE_SUPPORT_ENABLED`).
```

File: rfd/0058-desktop-file-transfer.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+### File System Virtual Channel Extension (RDP Option 2)
+
+From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all functionality on that directory is instigated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.
```
Comment (zmb3) [inline rfd/0058-desktop-file-transfer.md:25]:
```suggestion
From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all actions on that directory are initiated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.
```

Reworded this a bit, but it's still not quite clear to me what you're describing here.

File: rfd/0058-desktop-file-transfer.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+## Client Side Possibilities
+
+Irrespective of which option we choose, we will need some means of initiating and completing file transfers on the client side (from the browser).
+
+### Clipboard API
+
+At first glance, it seemed like using the `Clipboard.read` and `Clipboard.write` API's might be a possibile mechanism for us to make use of. However while some files can be transferred to the client's clipboard with this API, browsers are [constrained](https://w3c.github.io/clipboard-apis/#reading-from-clipboard) to the `text/plain` `text/html` and `image/png` mime types. Ergo unless we want to limit our users to a very narrow set of file transfers, this option is not available to us.
+
+### Browser Upload and Download (Client Option 1)
```
Comment (zmb3) [inline rfd/0058-desktop-file-transfer.md:41]: I'm having a hard time keeping track of which option is being discussed due to the deep level of nesting here.

Is "Client Option 1" a sub-option of RDP Option 2a?

**Review by zmb3 — COMMENTED** [review]
Minor request - could you reflow these super long lines so it's easier to comment?

File: rfd/0058-desktop-file-transfer.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+For example, if a program on the server wants to read a shared file, it sends a [`Device Read Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/3192516d-36a6-47c5-987a-55c214aa0441), and the client reads the requested data and sends it back with a [`Device Read Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/d35d3f91-fc5b-492b-80be-47f483ad1dc9). If instead the server-side program wants to write to a shared file, it sends a [`Device Write Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/2e25f0aa-a4ce-4ff3-ad62-ab6098280a3a) specifying where and what data it wants to write. In the happy path, the client writes that data to disk and sends back a [`Device Write Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/58160a47-2379-4c4a-a99d-24a1a666c02a) with the number of bytes written. If the client-side write fails, it sends back the `Device Write Response` with an `IoStatus` field set to something besides `STATUS_SUCCESS` (such as `STATUS_UNSUCCESSFUL`), which alerts the server make the write operation fail on its end as well.
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply rfd/0058-desktop-file-transfer.md:27]: I'm not 100% sure. I had the same question, and tested that it does get notified by experimenting with drive redirection in the official Microsoft Remote Desktop client. One possibility is that it polls, another is that it re-requests metadata (file/directory names and sizes) every time some program on the server side requests access to the shared drive.

File: rfd/0058-desktop-file-transfer.md (hunk 3 above)
Comment (ibeckermayer) [reply rfd/0058-desktop-file-transfer.md:41]: Yeah, my naming system is questionable, in hindsight the numbering might be more confusing than helpful. I tried to clarify by the structure of the headings, but its difficult to follow, particularly reading the plain Markdown. 

I've done some re-naming and here is a summary of the structure based on the headings:
- **RDP Possibilities and Data Flow (RDP Options 1, 2a, and 2b)**
  - Clipboard Virtual Channel Extension (RDP Option 1)
  - File System Virtual Channel Extension (RDP Option 2a and 2b)
- **Client Side Possibilities (Client Options 1 and 2)**
  - Browser Upload and Download (Client Option 1)
  - TDP Upload and Download via the File System Access API (Client Option 2)
- **Implementation Options (Full Implementation Options 1, 2, 3, and 4)**
  - RDP Option 1 + Client Option 1 (Full Implementation Option 1)
  - RDP Option 1 + Client Option 2 (Full Implementation Option 2)
  - RDP Option 2a + Client Option 1 (Full Implementation Option 3)
  - RDP Option 2b + Client Option 2 (Full Implementation Option 4)

The first two **bolded bullet points** contain our options on the backend and frontend respectively, the last one lays out the plausible combinations of first two that are worth considering.

File: rfd/0058-desktop-file-transfer.md [hunk 5]
```diff
@@ -0,0 +1,332 @@
+---
+authors: Isaiah Becker-Mayer (isaiah@goteleport.com)
+state: draft
+---
+
```
Comment (russjones) [inline rfd/0058-desktop-file-transfer.md:5]:
```suggestion
# Required Approvers
* Engineering: @zmb3 && (@probakowski || @LKozlowski)
* Product: (@klizhentas || @xinding33)
```

**Review by probakowski — COMMENTED** [review]
I wonder if we should restructure sections in a way to highlight option(s) that we will use instead of things we won't. Maybe move 2b to be 1a?

**Review by xinding33 — APPROVED** [review]
Option 2b provides the most intuitive and consistent user experience since uploading and downloading files follow the same flow. Additionally, many virtualization tools (e.g. VirtualBox) use the idea of a shared directory so it's likely a concept that our users will be familiar with.

**Review by klizhentas — APPROVED** [review]
2b lgtm

**Review by probakowski — APPROVED** [review]
2b LGTM

**Review by LKozlowski — APPROVED** [review]
Option 2B  - LGTM

As others have stated, and I agree with it, it's the most intuitive way to upload and download files.

**Pull request merged**